lib/pharos/ssh/client.rb

```diff
@@ -157,58 +46,54 @@ def connect
         @session = Net::SSH.start(@host, @user, @opts)
       end
 
+      # @example
+      #   tempfile do |tmp|
+      #     exec!("less #{tmp}")
+      #   end
+      # @example
+      #   tmp = tempfile.new(content: "hello")
+      #   exec!("cat #{tmp}")
+      #   tmp.unlink
+      #
+      # @param prefix [String] tempfile filename prefix (default "pharos")
+      # @param content [String,IO] initial file content, default blank
+      # @return [Pharos::SSH::Tempfile]
+      # @yield [Pharos::SSH::Tempfile]
+      def tempfile(prefix: "pharos", content: nil, &block)
+        Tempfile.new(self, prefix: prefix, content: content, &block)
+      end
+
       # @param cmd [String] command to execute
-      # @return [Exec]
+      # @return [Pharos::Command::Result]
       def exec(cmd, **options)
         require_session!
-
-        logger.debug { "exec: #{cmd}" }
-
-        ex = Exec.new(cmd, **options)
-        ex.open(@session)
-        ex.wait
-        ex
+        RemoteCommand.new(self, cmd, **options).run
       end
 
       # @param cmd [String] command to execute
-      # @raise [ExecError]
+      # @raise [Pharos::SSH::RemoteCommand::ExecError]
       # @return [String] stdout
       def exec!(cmd, **options)
-        ex = exec(cmd, **options)
-
-        raise ExecError.new(cmd, ex.exit_status, ex.output) if ex.error?
-
-        ex.stdout
+        require_session!
+        RemoteCommand.new(self, cmd, **options).run!.stdout
       end
 
       # @param script [String] name of script
       # @param path [String] real path to file, defaults to script
-      # @raise [ExecError]
+      # @raise [Pharos::SSH::RemoteCommand::ExecError]
       # @return [String] stdout
       def exec_script!(name, env: {}, path: nil, **options)
         script = File.read(path || name)
-        cmd = ['sudo']
-
-        env.each_pair do |e, value|
-          cmd << "#{e}=#{Shellwords.escape(value)}"
-        end
-
-        cmd << 'sh'
-        cmd << '-x'
-
-        ex = exec(cmd.join(' '), stdin: script, debug_source: name, **options)
-
-        raise ExecError.new(name, ex.exit_status, ex.output) if ex.error?
```

Comment:
One detail here that got lost in the refactoring: the `ExecError` contained the script name, not the `sh -x` command... now the resulting error unfortunately leaves it unclear which script failed to execute, although you can still figure it out from the backtrace.

Reply:
Hmm I'm sure I've seen errors like:

```
['sudo', 'sh', '-x'] < foofoo.sh
```

Need to check.